src/c_api/eigensolver/eigensolver.h

```diff
@@ -0,0 +1,154 @@
+//
+// Copyright (c) 2018-2023, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+#include <blas.hh>
+#include <mpi.h>
+
+#include <pika/init.hpp>
+
+#include <dlaf/common/assert.h>
+#include <dlaf/eigensolver/eigensolver.h>
+#include <dlaf/matrix/matrix.h>
+#include <dlaf/matrix/matrix_mirror.h>
+#include <dlaf/types.h>
+#include <dlaf_c/desc.h>
+#include <dlaf_c/grid.h>
+
+#include "../blacs.h"
+#include "../grid.h"
+
+template <typename T>
+int eigensolver(int dlaf_context, char uplo, T* a, DLAF_descriptor dlaf_desca, dlaf::BaseType<T>* w,
+                T* z, DLAF_descriptor dlaf_descz) {
+  try {
+    using MatrixHost = dlaf::matrix::Matrix<T, dlaf::Device::CPU>;
+    using MatrixMirror = dlaf::matrix::MatrixMirror<T, dlaf::Device::Default, dlaf::Device::CPU>;
+    using MatrixBaseMirror =
+        dlaf::matrix::MatrixMirror<dlaf::BaseType<T>, dlaf::Device::Default, dlaf::Device::CPU>;
+
+    DLAF_ASSERT(dlaf_desca.i == 0, dlaf_desca.i);
+    DLAF_ASSERT(dlaf_desca.j == 0, dlaf_desca.j);
+    DLAF_ASSERT(dlaf_descz.i == 0, dlaf_descz.i);
+    DLAF_ASSERT(dlaf_descz.j == 0, dlaf_descz.j);
+
+    pika::resume();
+
+    auto communicator_grid = dlaf_grids.at(dlaf_context);
+
+    dlaf::GlobalElementSize matrix_size(dlaf_desca.m, dlaf_desca.n);
+    dlaf::TileElementSize block_size(dlaf_desca.mb, dlaf_desca.nb);
+
+    dlaf::comm::Index2D src_rank_index(dlaf_desca.isrc, dlaf_desca.jsrc);
+
+    dlaf::matrix::Distribution distribution(matrix_size, block_size, communicator_grid.size(),
+                                            communicator_grid.rank(), src_rank_index);
+
+    dlaf::matrix::LayoutInfo layout_a = colMajorLayout(distribution, dlaf_desca.ld);
+    dlaf::matrix::LayoutInfo layout_z = colMajorLayout(distribution, dlaf_descz.ld);
+
+    MatrixHost matrix_host(distribution, layout_a, a);
+    MatrixHost eigenvectors_host(distribution, layout_z, z);
+    auto eigenvalues_host = dlaf::matrix::createMatrixFromColMajor<dlaf::Device::CPU>(
+        {dlaf_descz.m, 1}, {distribution.blockSize().rows(), 1}, std::max(dlaf_descz.m, 1), w);
+
+    {
+      MatrixMirror matrix(matrix_host);
+      MatrixMirror eigenvectors(eigenvectors_host);
+      MatrixBaseMirror eigenvalues(eigenvalues_host);
+
+      dlaf::eigensolver::eigensolver<dlaf::Backend::Default, dlaf::Device::Default, T>(
+          communicator_grid, blas::char2uplo(uplo), matrix.get(), eigenvalues.get(), eigenvectors.get());
+    }  // Destroy mirror
+
+    // Ensure data is copied back to the host
+    eigenvalues_host.waitLocalTiles();
+    eigenvectors_host.waitLocalTiles();
+
+    pika::suspend();
+    return 0;
+  }
+  catch (const std::exception& e) {
+    std::cerr << e.what() << '\n';
+    return -1;
+  }
+  catch (...) {
+    std::cerr << "ERROR: Unknown exception caught in DLA-Future's eigensolver." << '\n';
+    return -1;
+  }
+}
```

Comment:
I don't get what it is supposed to throw an exception that has to handled.
All the DLAF checks terminate if not fulfilled.